---
title: "Differential Gene Expression Analysis (DEA) Report: exercise 1"
author: "Mariangela Santorsola"
date: "15/11/2024"
output:
 html_document:
   toc: true
params:
    quant_path: /PATH/TO/salmon
    tx2gene: /PATH/TO/gencode.v29.transcripts_no-vers_chr21_tx2gene.txt
    
---

## Differential expression analysis

The goal of differential expression analysis is to compare gene expression levels between distinct experimental conditions, such as treatments and controls.

Determine which genes show statistically significant differences in expression between conditions.

Measure the magnitude and direction of changes in gene expression (e.g., upregulation or downregulation).

Understand the biological implications of expression changes, such as pathways, processes, or mechanisms influenced by treatments or conditions.


### Input data

**Note** Please double-check the file paths for all inputs referenced in this document.

- quant_path: Ensure that this params points to the directory "salmon" containing all individual sample folders:
```
.salmon                               # "salmon" directory created by nf-core/rnaseq 
├── **control_rep1**                  # Directory for an individual sample in cases/controls (e.g. cases, controls)
│   ├── aux_info
│   │   ├── ambig_info.tsv
│   │   ├── expected_bias.gz
│   │   ├── fld.gz
│   │   ├── meta_info.json
│   │   ├── observed_bias.gz
│   │   └── observed_bias_3p.gz
│   ├── cmd_info.json
│   ├── libParams
│   │   └── flenDist.txt
│   ├── lib_format_counts.json
│   ├── logs
│   │   └── salmon_quant.log
│   ├── quant.genes.sf
│   └── **quant.sf**
[...]
```
You can upload the compressed Salmon directory (e.g., salmon.zip) directly. RStudio will automatically uncompress it upon upload.

- tx2gene: ensure that you use the following file:: [gencode.v29.transcripts_no-vers_chr21_tx2gene.txt](https://github.com/lescai-teaching/datasets_reference_only/blob/main/trascriptome/gencode.v29.transcripts_no-vers_chr21_tx2gene.txt).


## R env setup

Load R libraries

```{r packages, eval=TRUE, echo=FALSE}
library(DESeq2)
library(tximport)
library(tidyverse)
library(pheatmap)
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
```


## Set conditions, e.g. cases/controls

```{r}
dataset <- tibble(
  sample = c("control_rep1",
             "control_rep2",
             "control_rep3",
             "treatment_rep1",
             "treatment_rep2",
             "treatment_rep3"),
  condition = c(rep("control", 3),
                rep("case", 3))
)

dataset
```



## Read the table linking transcripts to genes 

```{r}
tx2gene <- read_tsv(params$tx2gene, col_names = FALSE, show_col_types = FALSE)

```



## Read quantification files (*.quant) from salmon pseudo-alignment

```{r}
files <- file.path(params$quant_path, dataset$sample, "quant.sf")
names(files) <- dataset$sample
```


The function _tximport_ imports the necessary quantification data for DESeq2:

```{r}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
colnames(txi$counts)
rownames(dataset) <- colnames(txi$counts)

```


##  Construct a DESeqDataSet from the txi object and sample information in dataset

```{r}
dds <- DESeqDataSetFromTximport(txi, dataset, ~condition)
```



## Filter min counts >= 10

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```



## Set base level to "control"

```{r}
dds$condition <- relevel(dds$condition, ref = "control")
```



## Perform Differential Expression Analysis between cases and controls

```{r}
dds <- DESeq(dds)
```

DESeq2 main steps:

- estimating size factors
- using 'avgTxLength' from assays(dds), correcting for library size
- estimating dispersions
- gene-wise dispersion estimates
- mean-dispersion relationship
- final dispersion estimates
- fitting model and testing



##  Extract results table from the dds object

```{r}
res <- results(dds)
head(results(dds, tidy=TRUE)) #let's look at the results table 
```



## DE analysis summary

```{r}
summary(res)
```



### Order Genes by padj

```{r}
resOrdered <- res[order(res$padj),]
resOrdered
```


### Data visualization

#### MA plot

```{r maplot, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Plot showing the log2 fold change over the mean of normalized counts for analysed samples"}
plotMA(res, ylim=c(-3,3))
#abline(h=c(-1,1), col="dodgerblue", lwd=2)
```



## Dispersion plot

```{r}
pdf("plots_dispersion.pdf")
plotDispEsts(dds)
dev.off()
```




# Plot counts 

```{r}
pdf("plots_counts.pdf")
plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
dev.off() 

```



## Save results to file or RData

```{r}
resdata <- as_tibble(resOrdered)
resdata$gene <- rownames(resOrdered)
write_tsv(resdata, "deseq2_results.tsv")
save.image("deseq2_analysis.RData")
```



## CLUSTERING analysis

```{r}
ntd <- normTransform(dds)
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]

df <- as.data.frame(colData(dds)[,c("condition")])
```



### CLUSTERING visualization

```{r}
#pdf("plots_heatmap.pdf")
pheatmap(assay(ntd)[select,],
         cluster_cols=FALSE, annotation_col=df$condition)
#dev.off()
```



## Plot Principal Component Aanalysis
```{r}
pdf("plots_pca.pdf")
plotPCA(ntd, intgroup=c("condition"))
dev.off()
```



## Gene Ontology Enrichment Analysis

### Select the most significant genes by padj

```{r}
universe <- AnnotationDbi::select(org.Hs.eg.db,
                                  keys = keys(org.Hs.eg.db),
                                  columns = c('ENTREZID','SYMBOL','ENSEMBL','ENSEMBLTRANS'),
                                  keytype = 'ENTREZID')

sig_genes <- resdata$gene[which(resdata$padj<0.05)]
entrez_genes_sig <- unique(universe[which(universe$ENSEMBL %in% sig_genes),]$ENTREZID)

pvalue_ens_genes <- resdata$padj[which(resdata$padj<0.05)]
names(pvalue_ens_genes)<-sig_genes

pvalue_entrez_genes <- resdata$padj[which(resdata$padj<0.05)]
names(pvalue_entrez_genes) <- entrez_genes_sig

```



### check gene Ensembl ID / Gene symbol

```{r}
sig_genes_names <- tx2gene[which(tx2gene$gene_id %in% sig_genes),]
```



### GO Enrichment Analysis

```{r}
ego <- enrichGO( gene = sig_genes,
                 universe = unique(tx2gene$gene_id),
                 OrgDb = org.Hs.eg.db,
                 keyType = 'ENSEMBL',
                 ont = "CC", # alternatives are BP (biological processes) and MF (molecular functions). Check if you get significant results also with these alternatives
                 pAdjustMethod = "BH",
                 pvalueCutoff  = 0.05)

```



#### Dotplot

```{r}
#pdf("plots_ego.pdf")
dotplot(ego, showCategory=30)
#dev.off()
```


#### Gene-Concept Network

```{r}
#pdf("plots_network-ego.pdf")
cnetplot(ego, foldChange=resdata$log2FoldChange[which(resdata$padj<0.5)])
#dev.off()

```



### List of Differentially Expressed Genes

```{r significant_genes_table, eval=TRUE, echo=FALSE, include=TRUE}
resOrdered <- res[order(res$padj),]
resOrdered <- na.omit(resOrdered) ### omit "NA" values 
resOrdered$Gene <- rownames(resOrdered)
sign_de_genes_table <- as_tibble(resOrdered[resOrdered$padj < 0.1,])
sign_de_genes_table %>% 
    dplyr::select(Gene, log2FoldChange, pvalue, padj) %>%
    knitr::kable()
significant_de_genes <- sign_de_genes_table$Gene
```

