---
title: "Omics Data Analysis Report: datasets_LABOS2023"
author: "Mariangela Santorsola"
date: "03/01/2024"
output:
 html_document:
   toc: true
params:
    quant_path: /PATH/TO/UPSTREAM/SAMPLE_DIRECTORIES
    tx2gene: /PATH/TO/gencode.v29.transcripts_no-vers_chr21_tx2gene.txt
    vcf:  /PATH/TO/joint_variant_calling/joint_germline_recalibrated_snpEff.ann.vcf.gz
    ext_funcs: /PATH/TO/extract_annotations_full.R
    
---
**Important** Please double check the path of each file used as input by this document.
- quant_path: Here you need to include the path to directory upstream to the single samples, such as:
```
.salmon
├── **control_rep1**
│   ├── aux_info
│   │   ├── ambig_info.tsv
│   │   ├── expected_bias.gz
│   │   ├── fld.gz
│   │   ├── meta_info.json
│   │   ├── observed_bias.gz
│   │   └── observed_bias_3p.gz
│   ├── cmd_info.json
│   ├── libParams
│   │   └── flenDist.txt
│   ├── lib_format_counts.json
│   ├── logs
│   │   └── salmon_quant.log
│   ├── quant.genes.sf
│   └── **quant.sf**
[...]
```
- tx2gene: Make sure to use this [gencode.v29.transcripts_no-vers_chr21_tx2gene.txt](https://github.com/lescai-teaching/datasets_reference_only/blob/main/trascriptome/gencode.v29.transcripts_no-vers_chr21_tx2gene.txt) file.
- ext_funcs: Make sure to include in the path the external functions available in  to process VCF file [extract_annotations_full.R](https://github.com/santorsola-teaching/class-lab-adv-omics-2023/blob/157ec8c34ed0e51646224bf1a6ffd220b9356ea4/L34_from_mutations_to_proteins/extract_annotations_full.R)

# RNAseq data analysis

## Experimental design

```{r packages, eval=TRUE, echo=FALSE}
library(DESeq2)
library(tximport)
library(tidyverse)
library(pheatmap)
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
```

This section contains the results of the RNAseq data analysis.

```{r data_input, eval=TRUE, echo=FALSE, include=TRUE}
groups <- rep(c("control", "case"), each = 3)
samples <-c("sample_01", "sample_02", "sample_03", "sample_04", "sample_05", "sample_06")
data <- tibble(Group = groups, Sample = samples) 
data %>% 
    knitr::kable()
```

## Data Processing

### Read alignment and Quantification

[...]

```{bash, eval=FALSE, echo=TRUE, include=TRUE}
nextflow run nf-core/rnaseq -r 3.12.0 \
--input  datasets_lesson33/rnaseq/reads/rnaseq_samplesheet.csv \
--outdir results-in-you-bucket \
--genome GRCh38chr21 \
--pseudo_aligner salmon \
--skip_alignment \
--skip_biotype_qc \
-c rnaseq_nextflow.config \
-profile gls \
--skip_stringtie \
--skip_bigwig \
--skip_umi_extract \
--skip_trimming \
--skip_fastqc
```


- The config file used to launch nf-core/rnaseq is available at [rnaseq_nextflow.config](https://github.com/santorsola-teaching/class-lab-adv-omics-2023/blob/main/L34_from_mutations_to_proteins/rnaseq_nextflow.config)

## Differential Expression Analysis

[...]


```{r}
dataset <- tibble(
  sample = c("control_rep1",
             "control_rep2",
             "control_rep3",
             "treatment_rep1",
             "treatment_rep2",
             "treatment_rep3"),
  condition = c(rep("control", 3),
                rep("case", 3))
)


# Read transcript to gene ID table
tx2gene <- read_tsv(params$tx2gene, col_names = FALSE, show_col_types = FALSE)
#tx2gene$X2 <- paste(tx2gene$X2, tx2gene$X3, sep='_')


# Read quantification files (*.quant) from salmon pseudo-alignment
files <- file.path(params$quant_path, dataset$sample, "quant.sf")
names(files) <- dataset$sample

txi <- tximport(files, type = "salmon", tx2gene = tx2gene)

colnames(txi$counts)
rownames(dataset) <- colnames(txi$counts)


dds <- DESeqDataSetFromTximport(txi, dataset, ~condition)


# Filter min counts >= 10
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]


# set base level to "control"
dds$condition <- relevel(dds$condition, ref = "control")


# Perform Differential Expression Analysis between cases and controls

dds <- DESeq(dds)


# Extract results table from the dds object
res <- results(dds)
head(results(dds, tidy=TRUE)) #let's look at the results table 

summary(res)


### Top Genes by padj
resOrdered <- res[order(res$padj),]
```

### MA plot

```{r maplot, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Plot showing the log2 fold change over the mean of normalized counts for analysed samples"}
plotMA(res, ylim=c(-3,3))
abline(h=c(-1,1), col="dodgerblue", lwd=2)
```

[and all the plots you find useful for data investigation...]

### List of Differentially Expressed Genes

```{r significant_genes_table, eval=TRUE, echo=FALSE, include=TRUE}
resOrdered <- res[order(res$padj),]
resOrdered <- na.omit(resOrdered) ### omit "NA" values <---- 
resOrdered$Gene <- rownames(resOrdered)
sign_de_genes_table <- as_tibble(resOrdered[resOrdered$padj < 0.1,])
sign_de_genes_table %>% 
    dplyr::select(Gene, log2FoldChange, pvalue, padj) %>%
    knitr::kable()
significant_de_genes <- sign_de_genes_table$Gene
```



# Re-sequencing data analysis

## Experimental design

[...]

## Data Processing

### Variant calling and annotation

[...]


```{r variant_settings, echo=FALSE, include=FALSE}
library(knitr)
library(VariantAnnotation)
library(tidyverse)
library(reticulate)
#source("extract_annotations_full.R")
source(params$ext_funcs)
```

Read vcf file from nf-core/sarek:

```{r import_VCF, eval=TRUE, echo=TRUE, include=TRUE, warning=FALSE, message=FALSE}
vcf <- readVcf(params$vcf)
```

```{r rowRanges, eval=TRUE, echo=FALSE, include=FALSE}
head(rowRanges(vcf))
```

```{r infoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(info(vcf))
```

```{r genoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(geno(vcf)$GT)
```

Create the table of annotations by the INFO field within the VCF file:

```{r simplifyVCF, message=FALSE, warning=FALSE}
variants <- rowRanges(vcf)
variants$`REF` <- as.character(variants$`REF`)
variants$ALT <- sapply(variants$ALT, function(x) as.character(x)[1])
variants <- as_tibble(variants)
variants$variantName = names(rowRanges(vcf))
variants = cbind(variants, as_tibble(geno(vcf)$GT))
variants$gene <- unlist(lapply(info(vcf)$ANN, get_most_severe_gene))
variants$ens <- unlist(lapply(info(vcf)$ANN, get_most_severe_ens))
variants$transcript <- unlist(lapply(info(vcf)$ANN, get_most_severe_tr))
variants$aa_change <- unlist(lapply(info(vcf)$ANN, get_most_severe_aa_change))
variants$consequence <- unlist(lapply(info(vcf)$ANN, get_most_severe_consequence))
variants$impact <- unlist(lapply(info(vcf)$ANN, get_most_severe_impact))
```

### Localization and biological consequences of identified mutations

[...]

```{r consequences, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Distribution of the percentage of identified mutations across biological consequence categories"}
variants %>%
  filter(!is.na(consequence)) %>%
  count(consequence) %>%
  mutate(count = n,
         percent = paste0(round(count/sum(count) * 100, digits = 2), "%"))%>%
  arrange(desc(count)) %>%
  mutate(lab.ypos = cumsum(count) - 0.5 * count) %>%
  ggplot(aes(x="", y=count, fill=consequence))+
  geom_bar(width = 1, stat = "identity")+
  coord_polar(theta = "y")+
  ggrepel::geom_text_repel(aes(y=lab.ypos, label=percent), max.overlaps = Inf)+
  theme_void()
```

 or you can use a table format:


```{r consequences_table, eval=TRUE, echo=FALSE, include=TRUE}
variants %>%
  filter(!is.na(consequence)) %>%
  group_by(consequence) %>%
  summarise(count=n()) %>%
  arrange(desc(count)) %>%
  kable()
```


## Variant Filtering

### Filtering Criteria

[...]


The following *r* and *bash* code block are equivalent, as they output the filtered variants based on the defined criteria:


```{r filterVariants, eval=TRUE, echo=TRUE, include=TRUE}
filteredVars = variants %>%
  filter(control_control1 == "0/0" & 
            (impact == "MODERATE" | impact == "HIGH") &
            ens %in% significant_de_genes & 
            consequence == "missense_variant") 
filteredVars <- as_tibble(filteredVars)
```

or equivalent code in bash

```{bash, eval=FALSE}
zcat <vcf.gz> | perl -nae 'if($F[10]=~/0\/0/){print $_;}' | grep 'HIGH\|MODERATE' | grep 'GENE1\|GENE2\|GENE3\|GENE4' | grep missense > mutations_to_check_in_alphamissense.txt
```


### Filtered mutations for further validation assays

```{r filtered_vars_table, eval=TRUE, echo=FALSE, include=TRUE}
selectedVars <- filteredVars %>%
  dplyr::select(seqnames, start, REF, ALT, gene, consequence, aa_change) 

selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()

```

### Alphamissense Pathogenicity Predictions of filtered variants

You need to search the genes and related protein variants using the Alphamissense portal, you know from the related class.
```{r alphamissense, eval=TRUE, echo=FALSE, include=TRUE}
selectedVars$alphamis_Pathogenicity = c(0.085, 0.149, 0.977, 0.062, 0.076, 0.106)
selectedVars$alphamis_Class = c("likely_benign", "likely_benign", "likely_pathogenic", "likely_benign", "likely_benign", "likely_benign")

selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()
```

### Gene constraint metrics from gnomAD

You need to seard for LOEFF values related to investigated genes, to assess if...

```{r gene_contraint, eval=TRUE, echo=FALSE, include=TRUE}
genes <- c("SOD1", "ETS2", "PSMG1")
loeuf <-c(1.515, 0.434, 0.927)
data <- tibble(Gene = genes, LOEUF = loeuf) 
data %>% 
    knitr::kable()
```


## Results: Prioritized variant(s)

You can add more information about the prioritized variant(s), such as clinvar significance, variant allele frequency from gnomAD database, and all external tool and resources seen in the course... 


```{r prioritized_variants, eval=TRUE, echo=FALSE, include=TRUE}
prioritised_variants <- selectedVars %>% filter(alphamis_Class == "likely_pathogenic")
prioritised_variants$Clinical_significance <- "Pathogenic/Likely pathogenic"
prioritised_variants$gnomAD_AF <- "variant not found"
prioritised_variants %>% 
    knitr::kable()
```

