---
title: "resequencing and rnaseq data integration: exercise 1 datasets"
author: "Mariangela Santorsola"
date: "15/11/2024"
output:
 html_document:
   toc: true
params:
    quant_path: /Users/mariangelasantorsola/Documents/lab-adv-omics-2023/results_cloud_LABOS_2024/rnaseq/rnaseq_results_ex1/salmon_ex1
    tx2gene: /Users/mariangelasantorsola/Documents/lab-adv-omics-2023/datasets_LABOS_exercise3/results/salmon/gencode.v29.transcripts_no-vers_chr21_tx2gene.txt
    vcf:  /Users/mariangelasantorsola/Documents/lab-adv-omics-2023/results_cloud_LABOS_2024/sarek/sarek_results_ex1/annotation/haplotypecaller/joint_variant_calling/joint_germline_recalibrated_snpEff.ann.vcf.gz
    ext_funcs:  /Users/mariangelasantorsola/Downloads/extract_annotations_full.R
    
---

# Assignment goal


*Identification of Variants in Differentially Expressed Genes with Significant Clinical Relevance* 



## Differential gene expression analysis

Load R libraries:

```{r packages, eval=TRUE, echo=FALSE, message=FALSE, include=FALSE}
library(DESeq2)
library(tximport)
library(tidyverse)
library(pheatmap)
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
```


### Set conditions, e.g. cases/controls

```{r set_conditions, eval=TRUE, echo=TRUE}
dataset <- tibble(
  sample = c("control_rep1",
             "control_rep2",
             "control_rep3",
             "treatment_rep1",
             "treatment_rep2",
             "treatment_rep3"),
  condition = c(rep("control", 3),
                rep("case", 3))
)

dataset
```



### Read the table linking transcripts to genes 

```{r transc_gene_ids, eval=TRUE, echo=FALSE}
tx2gene <- read_tsv(params$tx2gene, col_names = FALSE, show_col_types = FALSE)

```

### Read quantification files (*.quant) from salmon pseudo-alignment

```{r load_quant_files, eval=TRUE, echo=FALSE}
files <- file.path(params$quant_path, dataset$sample, "quant.sf")
names(files) <- dataset$sample
```


The function _tximport_ imports the necessary quantification data for DESeq2:

```{r tximport_quant_data, eval=TRUE, echo=FALSE}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
colnames(txi$counts)
rownames(dataset) <- colnames(txi$counts)

```


###  Construct a DESeqDataSet from the txi object and sample information in dataset

```{r build_dds_object, eval=TRUE, echo=FALSE}
dds <- DESeqDataSetFromTximport(txi, dataset, ~condition)
```



### Filter min counts >= 10

```{r filter_low_exp_genes, eval=TRUE, echo=FALSE}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```


### Set base level to "control"

```{r set_baseline, eval=TRUE, echo=FALSE}
dds$condition <- relevel(dds$condition, ref = "control")
```


### Perform Differential Expression Analysis between cases and controls

```{r diff_expr_analysis, eval=TRUE, echo=FALSE}
dds <- DESeq(dds)
```

###  Extract results table from the dds object

```{r list_expr_genes, eval=TRUE, echo=FALSE}
res <- results(dds)
head(results(dds, tidy=TRUE)) #let's look at the results table 
```


### DE analysis summary

```{r DEA_summary, eval=TRUE, echo=FALSE}
summary(res)
```


#### Order Genes by padj

```{r sort_gene_list, eval=TRUE, echo=FALSE}
resOrdered <- res[order(res$padj),]
resOrdered
```


### List of Differentially Expressed Genes

```{r significant_genes_table, eval=TRUE, echo=FALSE, include=TRUE}
resOrdered <- res[order(res$padj),]
resOrdered <- na.omit(resOrdered) ### omit "NA" values 
resOrdered$Gene <- rownames(resOrdered)
sign_de_genes_table <- as_tibble(resOrdered[resOrdered$padj < 0.1,])
sign_de_genes_table %>% 
    dplyr::select(Gene, log2FoldChange, pvalue, padj) %>%
    knitr::kable()
significant_de_genes <- sign_de_genes_table$Gene
```




## Variant annotation, filtering and prioritization

Load R libraries:

```{r variant_settings, echo=FALSE, include=FALSE}
library(knitr)
library(VariantAnnotation)
library(tidyverse)
source(params$ext_funcs)
```


Read vcf file from nf-core/sarek:

```{r import_VCF, eval=TRUE, echo=TRUE, include=TRUE, warning=FALSE, message=FALSE}
vcf <- readVcf(params$vcf)
```

Check VCF file fields:

```{r rowRanges, eval=TRUE, echo=FALSE, include=FALSE}
head(rowRanges(vcf))
```


```{r infoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(info(vcf))
```


```{r genoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(geno(vcf)$GT)
```

### Create the table of annotations by the INFO field within the VCF file:

```{r simplifyVCF, message=FALSE, warning=FALSE}
variants <- rowRanges(vcf)
variants$`REF` <- as.character(variants$`REF`)
variants$ALT <- sapply(variants$ALT, function(x) as.character(x)[1])
variants <- as_tibble(variants)
variants$variantName = names(rowRanges(vcf))
variants = cbind(variants, as_tibble(geno(vcf)$GT))
names(variants)[names(variants) == "patient_01_sample_01"] <- "case_case1"
names(variants)[names(variants) == "patient_02_sample_02"] <- "control_control1"
variants$gene <- unlist(lapply(info(vcf)$ANN, get_most_severe_gene))
variants$ens <- unlist(lapply(info(vcf)$ANN, get_most_severe_ens))
variants$transcript <- unlist(lapply(info(vcf)$ANN, get_most_severe_tr))
variants$aa_change <- unlist(lapply(info(vcf)$ANN, get_most_severe_aa_change))
variants$consequence <- unlist(lapply(info(vcf)$ANN, get_most_severe_consequence))
variants$impact <- unlist(lapply(info(vcf)$ANN, get_most_severe_impact))
```


### Localization and biological consequences of identified mutations
 
```{r consequences_table, eval=TRUE, echo=FALSE, include=TRUE}
variants %>%
  filter(!is.na(consequence)) %>%
  group_by(consequence) %>%
  summarise(count=n()) %>%
  arrange(desc(count)) %>%
  kable()
```


## RNAseq and resequencing data integration

### Filtering Criteria

Filter:

- *missense* variants only *in the case sample* and NOT in the matched control
- variants featuring a "*MODERATE*" or "*HIGH*" impact
- variants occurring in differentially expressed (DE) genes in cases vs control samples


```{r filterVariants, eval=TRUE, echo=TRUE, include=TRUE}
filteredVars = variants %>%
  filter(control_control1 == "0/0" & 
            (impact == "MODERATE" | impact == "HIGH") &
             ens %in% significant_de_genes & 
            consequence == "missense_variant") 
filteredVars <- as_tibble(filteredVars)
```



Select relevant field within filteredVars table:

```{r filtered_vars_table, eval=TRUE, echo=FALSE, include=TRUE}
selectedVars <- filteredVars %>%
  dplyr::select(seqnames, start, REF, ALT, gene, consequence, aa_change) 
selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()

```

### Additional annotations

In addition to the Ensembl Variant Effect Predictor (VEP) tool (https://www.ensembl.org/Tools/VEP), you can mine the portal AlphaMissense (https://alphamissense.hegelab.org/) to retrieve pathogenicity predictions for missense variants.

You can add more information about the prioritized variant(s), such as clinvar significance (ClinVar), variant allele frequency (gnomAD), and all external tool and resources seen in the course.

### Gene constraint metrics from gnomAD

You need to search for LOEFF values related to investigated genes, to assess if...

```{r gene_contraint, eval=TRUE, echo=FALSE, include=TRUE}
genes <- c("SOD1", "ETS2", "PSMG1")
loeuf <-c(1.51, 0.434, 0.926)
data <- tibble(Gene = genes, LOEUF = loeuf) 
data %>% 
    knitr::kable()
```


#### Alphamissense

```{r}
selectedVars$am_pathogenicity = c(0.085, 0.149, 0.977, 0.062, 0.076, 0.106)
selectedVars$am_class = c("likely_benign", "likely_benign", "likely_pathogenic", "likely_benign", "likely_benign", "likely_benign")

```


#### ClinVar: clinical significance

```{r}
selectedVars$clinVar = c("Conflicting classifications of pathogenicity", "Conflicting classifications of pathogenicity", "Pathogenic/Likely pathogenic","Benign", "Benign","Benign")

```


#### gnomAD

```{r}
selectedVars$gnomAD_AF = c("7.50e-5","9.53e-4","4.97e-6","3.20e-3","9.33e-3", "5.35e-4")
#21-31659828-31659828

#7.50e-5 = 0.0000750

```


Table of annotated variants:

```{r}
selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()
```




## Prioritized variant(s)

```{r prioritized_variants, eval=TRUE, echo=FALSE, include=TRUE}
#seqnames	start	REF	ALT	gene	consequence	aa_change	am_pathogenicity	am_class	clinVar	gnomAD_AF
prioritised_variants <- selectedVars %>% 
    filter(am_class == "likely_pathogenic" | am_class == "Conflicting classifications of pathogenicity" &
               clinVar == "Pathogenic/Likely pathogenic")

prioritised_variants %>% 
    knitr::kable()
```




